src/Microsoft.OData.Core/Json/ODataJsonWriterUtilsAsync.cs

```diff
@@ -0,0 +1,274 @@
+            {
+                await jsonWriter.WritePaddingFunctionNameAsync(settings.JsonPCallback).ConfigureAwait(false);
+                await jsonWriter.StartPaddingFunctionScopeAsync().ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// If we are writing a response and the given Json Padding function name is not null or empty
+        /// this function will close the JSONP scope asynchronously.
+        /// </summary>
+        /// <param name="jsonWriter">JsonWriter to write to.</param>
+        /// <param name="settings">Writer settings.</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        internal static async Task EndJsonPaddingIfRequiredAsync(IJsonWriterAsync jsonWriter, ODataMessageWriterSettings settings)
+        {
+            Debug.Assert(jsonWriter != null, "jsonWriter should not be null");
+
+            if (settings.HasJsonPaddingFunction())
+            {
+                await jsonWriter.EndPaddingFunctionScopeAsync().ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously writes an error message.
+        /// </summary>
+        /// <param name="jsonWriter">JSON writer.</param>
+        /// <param name="code">The code of the error.</param>
+        /// <param name="message">The message of the error.</param>
+        /// <param name="target">The target of the error.</param>
+        /// <param name="details">The details of the error.</param>
+        /// <param name="innerError">Inner error details that will be included in debug mode (if present).</param>
+        /// <param name="instanceAnnotations">Instance annotations for this error.</param>
+        /// <param name="writeInstanceAnnotationsDelegate">Delegate to write the instance annotations.</param>
+        /// <param name="maxInnerErrorDepth">The maximum number of nested inner errors to allow.</param>
+        /// <param name="writingJsonLight">true if we're writing JSON lite, false if we're writing verbose JSON.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        private static async Task WriteErrorAsync(IJsonWriterAsync jsonWriter, string code, string message, string target,
+            IEnumerable<ODataErrorDetail> details,
+            ODataInnerError innerError,
+            IEnumerable<ODataInstanceAnnotation> instanceAnnotations,
+            Func<IEnumerable<ODataInstanceAnnotation>, Task> writeInstanceAnnotationsDelegate, int maxInnerErrorDepth,
+            bool writingJsonLight)
+        {
+            Debug.Assert(jsonWriter != null, "jsonWriter != null");
+            Debug.Assert(code != null, "code != null");
+            Debug.Assert(message != null, "message != null");
+            Debug.Assert(instanceAnnotations != null, "instanceAnnotations != null");
+
+            // "error": {
+            await jsonWriter.StartObjectScopeAsync().ConfigureAwait(false);

```

Comment:
>StartObjectScopeAsync [](start = 29, length = 21)

 I am wondering the performance since we have some many await in a function call/

Does each await create a thread? Could you do a research? Could we combine the write today and only call one await?

Reply:
Aside from eliding, maybe there are situations when they could be combined. Maybe in a future PR we could measure impact and see whether there are opportunities for improvements.